virtcontainers/api.go

```diff
@@ -64,6 +65,27 @@ func createSandboxFromConfig(sandboxConfig SandboxConfig) (*Sandbox, error) {
 		return nil, err
 	}
 
+	// rollback to stop VM if error occurs
+	defer func() {
+		if err != nil {
+			if sandboxRunning {
+				s.agent.stopSandbox(s)
+			}
+
+			s.stopVM()
+		}
+	}()
+
+	// Once startVM is done, we want to guarantee
+	// that the sandbox is manageable. For that we need
+	// to start the sandbox inside the VM.
+	if err = s.agent.startSandbox(s); err != nil {
+		return nil, err
+	}
+
+	// set the sandboxRunning state to true
+	sandboxRunning = true
```

Comment:
No need for this boolean if you define another defer function here:
```go
	defer func() {
		if err != nil {
			s.agent.stopSandbox(s)
		}
	}()
```
Because this defer function is defined after the first one, it will be called prior to it. That's a stack, not a FIFO, meaning it will do exactly what you need.
This way, no need for a complex defer function, just call a bunch of simple ones.

Reply:
Yeap, your suggestion is more graceful!